server/consumer.go

```diff
@@ -2070,8 +2075,10 @@ func (o *consumer) getNextMsg() (subj string, hdr, msg []byte, seq uint64, dc ui
 // Lock should be held.
 func (o *consumer) forceExpireFirstWaiting() *waitingRequest {
 	// FIXME(dlc) - Should we do advisory here as well?
-	wr := o.waiting.pop()
-	if wr == nil {
+	wr := o.waiting.peek()
```

Comment:
This seems at first glance to be same logic as pop, so what is different here?

Reply:
Here we are removing the fetch request forcefully regardless of the batch size, while `pop()` keeps the request valid until the batch request is fulfilled or if there is still interest. This is to avoid filling the queue with stale requests for example if the queue is filled with requests like `sub.Fetch(1000, nats.MaxWait(1*time.Hour)`, then no new fetch requests can be made for the consumer until the 1000 elements are received or one of the request expires.
With this change, in case of reaching the num waiting limit of inflight fetch requests, the older fetch requests are force expired to give room to the newer fetch requests.

Reply:
We have code that is supposed to deal with requests becoming invalid or expiring, why does that code not work here?